LMNT TTS

sharvil opened the pull request.

sharvil commented:
@chadbailey59, here's an initial cut of adding LMNT as a `TTSService`. Happy to revisit https://github.com/pipecat-ai/pipecat/pull/87 or carry on the work here.

aconchillo commented:
Awesome. Thank you! Is there any way to know when the audio is done? We use `TTSStopFrame` for later generating `BotStoppedSpeakingFrame`. This is a common problem with streaming TTS. Most services use timing, that is, if no text has been received in a short period of time it means we are probably done.

aconchillo commented on pyproject.toml:
```diff
@@ -45,6 +45,7 @@ gladia = [ "websockets~=12.0" ]
 google = [ "google-generativeai~=0.7.1" ]
 fireworks = [ "openai~=1.35.0" ]
 langchain = [ "langchain~=0.2.10", "langchain-community~=0.2.9", "langchain-openai~=0.1.17" ]
+lmnt = [ "lmnt~=1.1.4" ]
```
There's a conflict in this file. 🙏

sharvil commented:
We have a `buffer_empty` message when the server buffer is drained (i.e. it has produced speech for all of the text sent up so far). There's no server-side delay so it could be that the server responds with a `buffer_empty` even though there's text in-flight to the server over the network.

We have a couple of options:
1. if the semantics of `TTSStoppedFrame` are such that the rest of the pipeline handles spurious frames gracefully (`AudioRawFrame` -> ... -> `TTSStoppedFrame` -> `TTSStartedFrame` -> ...) then I can use the `buffer_empty` signal.
2. if not, I can add a client-side timer task to fire off the `TTSStoppedFrame` signal.

WDYT?

aconchillo commented:
> We have a `buffer_empty` message when the server buffer is drained (i.e. it has produced speech for all of the text sent up so far). There's no server-side delay so it could be that the server responds with a `buffer_empty` even though there's text in-flight to the server over the network.
> 
> We have a couple of options:
> 
> 1. if the semantics of `TTSStoppedFrame` are such that the rest of the pipeline handles spurious frames gracefully (`AudioRawFrame` -> ... -> `TTSStoppedFrame` -> `TTSStartedFrame` -> ...) then I can use the `buffer_empty` signal.
> 2. if not, I can add a client-side timer task to fire off the `TTSStoppedFrame` signal. Do you have a recommendation on how long the timeout should be here?
> 
> WDYT?

Since this the bot talking, the safest one is probably (2) plus it could be useful for other services which don't provide `buffer_empty` (or similar). I'm worried that option (1) might generate too many bot started/stopped frames when there's no need to.

About the timeout, I believe 800ms is a good number. I would take care of making this generic for other services.

Also... Any chance you could add an example to `examples/foundation` (07X-interruptible....).

sharvil commented:
I've added a generic mechanism to (optionally) produce `TTSStoppedFrame`s after a configurable timeout. Please let me know if this is roughly what you had in mind.

I'll add an example to this PR in my next commit.

aconchillo commented on src/pipecat/services/ai_services.py:
```diff
@@ -227,6 +237,22 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+        if self._stop_frame_task is not None:
+            # Reschedule timeout task if it was already running
+            self._stop_frame_task.cancel()
+            self._stop_frame_task = self.get_event_loop().create_task(self._stop_frame_handler())
```
I believe this will generate lots of task creation. There are services that produce thousands of `AudioRawFrame`s. Also, we would only do this if `isinatance(frame, AudioRawFrame)`.

Just talking out loud, but I was thinking that maybe we could just use one task and a queue with a `async.wait_for(queue.get(), timeout)`.

Something like this

```
if isinstance(frame, StartInterruptionFrame) or isinstance(frame, TTSStartedFrame) or  isinstance(frame, AudioRawFrame):
  await self._tts_stop_queue.put(frame)
```

then

```
async def _stop_frame_handler(self):
    tts_start_frame_seen = False
    while True:
        try:
            frame = await asyncio.wait_for(self._tts_stop_queue.get(), self._stop_frame_timeout)
            if isinstance(frame, StartInterruptionFrame):
                    tts_start_frame_seen = False
            elif isinstance(frame, TTSStartFrame):
                    tts_start_frame_seen = True
        except asyncio.TimeoutError:
            if tts_start_frame_seen:
                await self.push_frame(TTSStoppedFrame())
                tts_start_frame_seen = False
```

Maybe? Totally untested...

aconchillo commented on src/pipecat/services/ai_services.py:
```diff
@@ -227,6 +237,22 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+
+    async def _stop_frame_handler(self):
+        await sleep(self._stop_frame_timeout_s)
+        self.push_frame(TTSStoppedFrame())
```
This is missing an `await`

sharvil commented:
Thanks for the feedback! I updated the original approach to fix the missing `await` and only reschedule the task on `RawAudioFrame`. I also took your suggestion and added a commit to the PR to use a single long-lived task. Either approach works for me, let me know which one you prefer and we'll go with that.

Out of curiosity, I took some measurements to understand the baseline cost of task creation / task memory. It feels quite lightweight – about 8 microseconds to create a task on my Macbook, and about 1kB of memory per task. Since we only keep one task around at any time, the memory overhead doesn't really matter. On run time, we'd only notice a meaningful (~1ms) cost if a service is producing ~100 `RawAudioFrame`s per second.

**Task creation time**
https://gist.github.com/sharvil/775ba412f120f4c48cdabd04a5261cb7

**Memory use per task**
https://gist.github.com/sharvil/d1b6240d46a9433d88284e521a4eda14

aconchillo commented on src/pipecat/services/ai_services.py:
```diff
@@ -227,6 +238,33 @@ async def process_frame(self, frame: Frame, direction: FrameDirection):
+                isinstance(frame, AudioRawFrame)):
+            if self._stop_frame_task is None:
+                event_loop = self.get_event_loop()
+                self._stop_frame_task = event_loop.create_task(self._stop_frame_handler())
```
Maybe we can create the task in the constructor `if `self._push_stop_frames` is true.

aconchillo replied:
One thing we are missing is:
```
    async def stop(self, frame: EndFrame):
        if self._stop_frame_task: 
            self._stop_frame_task.cancel()
            await self._stop_frame_task
    async def cancel(self, frame: CancelFrame):
        if self._stop_frame_task: 
            self._stop_frame_task.cancel()
            await self._stop_frame_task
```

Otherwise when the bot stops it will hang there waiting for this task to finish.

sharvil replied:
I considered creating the task in the constructor but that would break the current pattern pipecat uses to override the base class' attributes. The current approach is to let the base class' constructor run to completion and *then* override the attributes (see https://github.com/pipecat-ai/pipecat/blob/main/src/pipecat/services/cartesia.py#L59-L69). If we created the task based on the value passed to the constructor, overriding the `_push_stop_frames` attribute in the derived class' constructor wouldn't do anything.

I added the task cancellation bits, lmk how you'd want to handle the constructor issue.

aconchillo replied:
Oh, I see. Yes, I guess that's specific to Cartesia's implementation. But yes, you are right I forgot you ended up making this generic.

Also, I forgot to mention, that if not the constructor it should probably be added to `start()`. I believe that makes more sense.

```
    async def start(self, frame: StartFrame):
        if self._push_stop_frames:
            event_loop = self.get_event_loop()
            self._stop_frame_task = event_loop.create_task(...)
```

sharvil replied:
Makes sense – moved task creation to `start`

aconchillo commented:
> Thanks for the feedback! I updated the original approach to fix the missing `await` and only reschedule the task on `RawAudioFrame`. I also took your suggestion and added a commit to the PR to use a single long-lived task. Either approach works for me, let me know which one you prefer and we'll go with that.
> 
> Out of curiosity, I took some measurements to understand the baseline cost of task creation / task memory. It feels quite lightweight – about 8 microseconds to create a task on my Macbook, and about 1kB of memory per task. Since we only keep one task around at any time, the memory overhead doesn't really matter. On run time, we'd only notice a meaningful (~1ms) cost if a service is producing ~100 `RawAudioFrame`s per second.
> 
> **Task creation time** https://gist.github.com/sharvil/775ba412f120f4c48cdabd04a5261cb7
> 
> **Memory use per task** https://gist.github.com/sharvil/d1b6240d46a9433d88284e521a4eda14

This is great! Thank you so much for taking the extra time to write this. I believe it's probably better to just have a long-lived task and a queue, it feels a bit simpler and a more common pattern in general.

Just added a couple more comments. I'll try it out after that. Thanks again!!!

aconchillo commented:
LGTM! Thank you!
